wxGrid: minor fixups and cleanups

AliKet opened the pull request:
This is the first PR in an attempt to implement this feature request (#22694)

In other words, i want to add transparent selection to `wxGrid` , keeping the old mechanism
in place as alternative/user preference (or as a fallback where transparency is not supported by the target system)

**Old:**
![Screenshot_2022-09-12_14-42-56](https://user-images.githubusercontent.com/7704771/189698497-e5a24cea-d8ab-4437-9467-c8a6bf2feb07.png)

**New:**
![Screenshot_2022-09-12_14-43-19](https://user-images.githubusercontent.com/7704771/189698719-e1bffc13-4080-4de4-abdc-64139ca3da76.png)

Commit:
Use wxEVT_NULL (not false) in SelectBlockNoEvent()

Commit:
Cleanup!

Use the already defined wxGridBlockCoordsVector type (defined in grid.h)
instead of introducing a new one, i.e. wxVectorGridBlockCoords

Commit:
Make wxGridSelection::GetBlocks() totally const

Changing or altering m_selection outside wxGridSelection itself will never be allowed

Commit:
Avoid generating deselection events when there is no selection

Each time the grid cursor moves it calls wxGrid::ClearSelection()
which calls wxGridSelection::ClearSelection(). the latter generates
deselection event which makes no sense if there is no selection at all.

AliKet commented:
Any objections to implement `wxGrid::SetCurrentCell()` uniformly under all platforms like this:

```
diff --git a/src/generic/grid.cpp b/src/generic/grid.cpp
index 0bcaaa7d5e..e07ed80eec 100644
--- a/src/generic/grid.cpp
+++ b/src/generic/grid.cpp
@@ -6436,51 +6436,13 @@ bool wxGrid::SetCurrentCell( const wxGridCellCoords& coords )
 
         if ( IsVisible( m_currentCellCoords, false ) )
         {
-#if defined(__WXOSX__) || defined(__WXGTK3__)
             RefreshBlock(m_currentCellCoords, m_currentCellCoords);
-#else
-            wxGridWindow* prevGridWindow = CellToGridWindow(m_currentCellCoords);
-            wxRect r;
-            r = BlockToDeviceRect( m_currentCellCoords, m_currentCellCoords, prevGridWindow );
-            if ( !m_gridLinesEnabled )
-            {
-                r.x--;
-                r.y--;
-                r.width++;
-                r.height++;
-            }
-
-            wxGridCellCoordsArray cells = CalcCellsExposed( r, prevGridWindow );
-
-            // Otherwise refresh redraws the highlight!
-            m_currentCellCoords = coords;
-
-            wxClientDC prevDc( prevGridWindow );
-            PrepareDCFor(prevDc, prevGridWindow);
-
-            DrawGridCellArea( prevDc, cells );
-            DrawAllGridWindowLines( prevDc, r , prevGridWindow);
-
-            if ( prevGridWindow->GetType() != wxGridWindow::wxGridWindowNormal )
-                DrawFrozenBorder(prevDc, prevGridWindow);
-#endif
         }
     }
 
     m_currentCellCoords = coords;
 
-#if defined(__WXOSX__) || defined(__WXGTK3__)
     RefreshBlock(coords, coords);
-#else
-    if ( ShouldRefresh() )
-    {
-        wxGridCellAttrPtr attr = GetCellAttrPtr( coords );
-        wxGridWindow* currentGridWindow = CellToGridWindow(coords);
-        wxClientDC dc( currentGridWindow );
-        PrepareDCFor(dc, currentGridWindow);
-        DrawCellHighlight( dc, attr.get() );
-    }
-#endif
 
     return true;
 }
```
IMHO, i don't see much gain in drawing the grid cursor from outside the `wxEVT_PAINT` event instead of just refreshing the affected rectangle. this makes `wxGrid::SetCurrentCell()` implementation much simpler, especially when adding label-selection/highlight latter ?

Commit:
Fix testing range-selection

The previous commit reveals a bug in the grid test where the
wxEVT_GRID_RANGE_SELECTED detected is not really comming from
range selection action, but from the grid cursor changing its position

For unknown reasons, range selection in the test doesn't really starts off
unless we move the mouse first while still inside the first selected cell

AliKet commented:
Not sure why this is still failing even if it succeeds locally ?

vadz commented on include/wx/generic/gridsel.h:
```diff
@@ -19,7 +19,6 @@
 
 #include "wx/vector.h"
 
-typedef wxVector<wxGridBlockCoords> wxVectorGridBlockCoords;
```
Not sure how did we end up with duplicate `wxGridBlockCoordsVector` and `wxVectorGridBlockCoords` typedefs, but unfortunately we can't just remove it from the public header any longer, even if it's clearly not needed.

Instead, it should be deprecated, with the message saying to use `wxGridBlockCoordsVector` instead, but left here (in a separate commit).

vadz reviewed (commented):
Thanks, this looks good except for removing the existing typedef. Please let me know if you have any problems with deprecating it.

vadz commented:
> Not sure why this still fails even though it succeeds locally ?

Unfortunately I don't know neither, but if the failure is reproducible, we need to fix it before merging this as we don't want to break the CI builds.

To debug it locally, please try running the test inside Xvfb.

vadz commented:
> Any objections to implement `wxGrid::SetCurrentCell()` uniformly under all platforms like this:

No, I don't think we really lose anything even under MSW by doing it like this. If you can test it under MSW, please still check that this doesn't produce any noticeable flicker when moving the cursor in the grid (and if you can't, please let me know so that I could do this).

Commit:
Skip test known to fail under GTK2 in automatic mode

Grid::RangeSelect, Grid::Size, Grid::ResizeScrolledHeader and Grid::ColumnMinWidth
are known to fail under GTK2 in automatic mode for unknown reason

Commit:
Restore wxVectorGridBlockCoords typedef removed in commit 40d3649d2e

The typedef should be kept for compatibility reasons, but with a
deprecation message to use wxGridBlockCoordsVector instead

AliKet commented:
> > Any objections to implement `wxGrid::SetCurrentCell()` uniformly under all platforms like this:
> 
> No, I don't think we really lose anything even under MSW by doing it like this. If you can test it under MSW, please still check that this doesn't produce any noticeable flicker when moving the cursor in the grid (and if you can't, please let me know so that I could do this).

In fact, I tested it before asking and saw no difference (i.e. no flickering and no loss in performance either)

vadz commented on tests/controls/gridtest.cpp:
```diff
@@ -659,6 +659,12 @@ TEST_CASE_METHOD(GridTestCase, "Grid::RangeSelect", "[grid]")
         return;
 
+#ifdef __WXGTK20__
+    // Works locally, but not when run on Travis CI.
```
"Travis CI" is a copy/paste, should be "GitHub CI", right?

AliKet replied:
Oops! will correct this and the others in a moment, thanks.

vadz commented:
> @vadz is it possible to rerun just the failing test because i don't see how is it related to this PR ? TIA !

I don't see neither but it definitely looks like it is related because I don't remember this check failing in the past and it seems to consistently fail now. I'd rather like to keep running all the tests, we spent much time on fixing them and it would be a pity to undo this work, even partially.

Note that the current check disables the test for both GTK 2 and 3, I don't care that much about GTK 2 any longer, so maybe we could live with disabling it for it (although it would still be pretty weird that it behaves differently from GTK 3), but I'd really like to keep the test with GTK 3.

Have you already tried running it under Xvfb locally to check if it fails there?

AliKet commented:
> Note that the current check disables the test for both GTK 2 and 3, I don't care that much about GTK 2 any longer, so maybe we could live with disabling it for it (although it would still be pretty weird that it behaves differently from GTK 3), but I'd really like to keep the test with GTK 3.
> 

This is embarrassing because the other tests (i.e. `Grid::Size, Grid::ResizeScrolledHeader` and `Grid::ColumnMinWidth`) actually pass on GitHub CI and the erroneous check for GTK 2 also disables  GTK 3! my bad!

> Have you already tried running it under Xvfb locally to check if it fails there?

I run the test locally **with and without** Xvfb and both GTK 2&3 pass successfully.

Commit:
Don't disable automated tests which work under GTK 3

Those tests are supposed to be skipped under GTK 2 only

Commit:
Disable test not working under GitHub CI using ASan

AliKet commented:
Now i don't know what should i do with this failure at all because running the test with ASan under Ubuntu 18.04 succeeds with and without Xvfb. and i don't really know how to debug something working for me in a local machine !!

AliKet commented:
I've recompiled everything with the exact configuration options as the one using ASAN
but under Ubuntu 18.04 (GTK 3.22.30) and here is the output of the run:

```
$ xvfb-run -a -s '-screen 0 1600x1200x24' ./test_gui [grid]
Test program for wxWidgets GUI features
build: 3.3.0 (wchar_t,compiler with C++ ABI compatible with gcc 4,wx containers,compatible with 3.2)
compiled using gcc 8.4
running under Linux 4.15.0-192-generic i686 as aliket
===============================================================================
All tests passed (441 assertions in 41 test cases)

```

The test also succeeds if i run the whole testsuit (just i had to disable ImageTestCase which raise an exception and crash) except this one:
```
ChoiceTestCase
  GetBestSize
-------------------------------------------------------------------------------
../../tests/controls/choicetest.cpp:36
...............................................................................

../../tests/controls/choicetest.cpp:117: FAILED:
  CHECK( bestSize.GetHeight() < m_choice->FromDIP(35) )
with expansion:
  35 < 35
```

AliKet commented:
I think now the following test cases (`Grid::Size, Grid::ResizeScrolledHeader, Grid::ColumnMinWidth and Grid::RangeSelect`) are really covered by the tests on GitHub CI. i.e. : 
- The last one really detects the event from the range-selection and not from cell-selection (or the position of the cursor changes)
- The others were erroneousslly skipped under GTK3

vadz commented:
Thanks, I think this is probably good enough, even if it would still be interesting to know what's going on in GitHub environment (I hadn't thought about it, but maybe using tmate and ssh'ing into the failing build might give more information...), so I'm going to merge it.

Thanks again!
